File: libs/document-processor/src/merge/processors/__tests__/ExpressionPreserver.test.ts
import { ExpressionPreserver } from '../ExpressionPreserver';
import {
  Blue,
  BasicNodeProvider,
  MergingProcessors,
} from '@blue-labs/language';

describe('ExpressionPreserver', () => {
  let nodeProvider: BasicNodeProvider;
  let blue: Blue;

  beforeEach(() => {
    nodeProvider = new BasicNodeProvider();

    // Set up test types
    nodeProvider.addSingleDocs(`
name: Json Patch Entry
val:
  description: The value to be used in the operation`);

    nodeProvider.addSingleDocs(`
name: Update Document
changeset:
  type: List
  itemType:
    blueId: ${nodeProvider.getBlueIdByName('Json Patch Entry')}`);

    // Create merger with ExpressionPreserver
    const mergingProcessor = new MergingProcessors.SequentialMergingProcessor([
      new MergingProcessors.ValuePropagator(),
      new ExpressionPreserver(),
      new MergingProcessors.TypeAssigner(),
    ]);

    blue = new Blue({ nodeProvider, mergingProcessor });
  });

  test('preserves expression values when type resolution would override them', () => {
    const yaml = `
type:
  blueId: ${nodeProvider.getBlueIdByName('Update Document')}
changeset: "\${steps.CreateSubscriptions.changes}"`;

    const source = blue.yamlToNode(yaml);
    const result = blue.resolve(source);

    // Expression should be preserved despite type saying changeset is a List
    expect(result.getProperties()?.['changeset']?.getValue()).toBe(
      '${steps.CreateSubscriptions.changes}',
    );
  });

  test('preserves nested property expressions', () => {
    nodeProvider.addSingleDocs(`
name: Complex Type
config:
  database:
    host:
      type: Text
    port:
      type: Integer`);

    const yaml = `
type:
  blueId: ${nodeProvider.getBlueIdByName('Complex Type')}
config:
  database:
    host: "\${env.DB_HOST}"
    port: "\${env.DB_PORT}"`;

    const source = blue.yamlToNode(yaml);
    const result = blue.resolve(source);

    // Both expressions should be preserved
    const db = result.getProperties()?.['config']?.getProperties()?.[
      'database'
    ];
    expect(db?.getProperties()?.['host']?.getValue()).toBe('${env.DB_HOST}');
    expect(db?.getProperties()?.['port']?.getValue()).toBe('${env.DB_PORT}');
  });

  test('does not affect properties without expressions', () => {
    const yaml = `
type:
  blueId: ${nodeProvider.getBlueIdByName('Update Document')}
changeset:
  - val: "regular value"`;

    const source = blue.yamlToNode(yaml);
    const result = blue.resolve(source);

    // Regular values should be processed normally
    const changeset = result.getProperties()?.['changeset'];
    expect(changeset?.getItems()).toBeDefined();
    expect(
      changeset?.getItems()?.[0]?.getProperties()?.['val']?.getValue(),
    ).toBe('regular value');
  });
});

File: references/java-sdk/src/test/java/blue/language/merge/ExpressionPreserverTest.java
package blue.language.merge;

import blue.language.Blue;
import blue.language.model.Node;
import blue.language.provider.BasicNodeProvider;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;

class ExpressionPreserverTest {

    @Test
    void preservesExpressionsWhenTypeResolutionWouldCoerceValueShape() {
        BasicNodeProvider nodeProvider = new BasicNodeProvider();
        nodeProvider.addSingleDocs(
                "name: Json Patch Entry\n" +
                        "val:\n" +
                        "  description: The value to be used in the operation");
        String patchEntryBlueId = nodeProvider.getBlueIdByName("Json Patch Entry");
        nodeProvider.addSingleDocs(
                "name: Update Document\n" +
                        "changeset:\n" +
                        "  type: List\n" +
                        "  itemType:\n" +
                        "    blueId: " + patchEntryBlueId
        );
        Blue blue = new Blue(nodeProvider);

        Node source = blue.yamlToNode("type:\n" +
                "  blueId: " + nodeProvider.getBlueIdByName("Update Document") + "\n" +
                "changeset: \"${steps.CreateSubscriptions.changes}\"");

        Node resolved = blue.resolve(source);
        Node changeset = resolved.getProperties().get("changeset");
        assertNotNull(changeset);
        assertEquals("${steps.CreateSubscriptions.changes}", changeset.getValue());
    }

    @Test
    void preservesNestedExpressionsAcrossObjectProperties() {
        BasicNodeProvider nodeProvider = new BasicNodeProvider();
        nodeProvider.addSingleDocs(
                "name: Complex Type\n" +
                        "config:\n" +
                        "  database:\n" +
                        "    host:\n" +
                        "      type: Text\n" +
                        "    port:\n" +
                        "      type: Integer"
        );
        Blue blue = new Blue(nodeProvider);

        Node source = blue.yamlToNode("type:\n" +
                "  blueId: " + nodeProvider.getBlueIdByName("Complex Type") + "\n" +
                "config:\n" +
                "  database:\n" +
                "    host: \"${env.DB_HOST}\"\n" +
                "    port: \"${env.DB_PORT}\"");
        Node resolved = blue.resolve(source);

        Node database = resolved.getProperties().get("config")
                .getProperties().get("database");
        assertEquals("${env.DB_HOST}", database.getProperties().get("host").getValue());
        assertEquals("${env.DB_PORT}", database.getProperties().get("port").getValue());
    }

    @Test
    void doesNotAlterRegularValuesWithoutExpressions() {
        BasicNodeProvider nodeProvider = new BasicNodeProvider();
        nodeProvider.addSingleDocs(
                "name: Json Patch Entry\n" +
                        "val:\n" +
                        "  description: The value to be used in the operation");
        String patchEntryBlueId = nodeProvider.getBlueIdByName("Json Patch Entry");
        nodeProvider.addSingleDocs(
                "name: Update Document\n" +
                        "changeset:\n" +
                        "  type: List\n" +
                        "  itemType:\n" +
                        "    blueId: " + patchEntryBlueId
        );
        Blue blue = new Blue(nodeProvider);

        Node source = blue.yamlToNode("type:\n" +
                "  blueId: " + nodeProvider.getBlueIdByName("Update Document") + "\n" +
                "changeset:\n" +
                "  - val: regular value");

        Node resolved = blue.resolve(source);
        Node changeset = resolved.getProperties().get("changeset");
        assertNotNull(changeset);
        assertNotNull(changeset.getItems());
        assertEquals(1, changeset.getItems().size());
        Node item = changeset.getItems().get(0);
        assertNotNull(item.getProperties());
        assertNotNull(item.getProperties().get("val"));
        assertEquals("regular value", item.getProperties().get("val").getValue());
        assertNull(item.getProperties().get("type"));
    }
}
